src/linux/gpio.c

```diff
@@ -0,0 +1,134 @@
+// Very basic support via a Linux gpiod device
+//
+// Copyright (C) 2017-2018  Kevin O'Connor <kevin@koconnor.net>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+#include "autoconf.h"
+#include "gpio.h"
+#include "command.h" // shutdown
+#include "sched.h" // shutdown
+
+#include <fcntl.h> // open
+#include <stdio.h> // snprintf
+#include <string.h> // memset
+#include <stdlib.h> // atexit
+
+#include "internal.h" // report_errno
+#include <gpiod.h>
+
+
+static struct gpiod_chip * chip = NULL;
+#define CHIP_NAME "gpiochip0"
+
+DECL_ENUMERATION_RANGE("pin", "P0", 0, 40);
+
+static void
+close_chip(void)
+{
+    gpiod_chip_close(chip);
+}
+
+static struct gpiod_chip*
+get_chip(void) {
+    if (chip == NULL) {
+        chip = gpiod_chip_open_by_name(CHIP_NAME);
+        if (chip == NULL) {
+            report_errno("gpiod_chip_open_by_name",-1);
+            shutdown("Unable to open GPIO chip device");
+        }
+        atexit(&close_chip);
+    }
+    return chip;
+}
+
+// Dummy versions of gpio_out functions
+struct gpio_out
+gpio_out_setup(uint32_t pin, uint8_t val)
+{
+    int ret;
+    struct gpiod_line* line = gpiod_chip_get_line(get_chip(), pin);
+    if (line == NULL) {
+        report_errno("gpiod_chip_get_line",-1);
+        shutdown("Unable to open GPIO chip line");
+    }
+    if (gpiod_line_is_requested(line)) {
+        gpiod_line_release(line);
+    }
+    if ((ret = gpiod_line_request_output(line,"klipper",val)) != 0) {
+        report_errno("gpiod_line_request_output",ret);
+        shutdown("Unable to request GPIO line output");
+    }
+    gpiod_line_set_value(line,val?1:0);
+    struct gpio_out g = { .line = line };
+    return g;
+}
+
+void
+gpio_out_reset(struct gpio_out g, uint8_t val)
+{
+    int ret;
+    if (gpiod_line_is_requested(g.line)) {
+        gpiod_line_release(g.line);
+    }
+    if ((ret = gpiod_line_request_output(g.line,"klipper",val)) != 0) {
+        report_errno("gpiod_line_request_output",ret);
+        shutdown("Unable to reset GPIO line output");
+    }
+}
+
+void
+gpio_out_write(struct gpio_out g, uint8_t val)
+{
+    gpiod_line_set_value(g.line,val?1:0);
+}
+
+void
+gpio_out_toggle(struct gpio_out g)
+{
+    gpiod_line_set_value(g.line,!gpiod_line_get_value(g.line));
```

Comment:
Thanks.  Have you verified that `gpiod_line_get_value()` here returns the last written value and not the sensed value of the gpio line?  The latter could lead to havoc.

-Kevin

Reply:
Hm, I was sure that this is worked as expected. But viewing source get me doubt. gpiod_line_get_value directly read register, some as for input line. It allowed for output line, but result is dependent on platform and implementation. You are right, It need be rewritten with local cache last written value.